### PR TITLE
HAAR-871 add group to user

### PR DESCRIPTION
**Pull request by AlTurner-MOJ**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserGroupController.kt [hunk 1]
```diff
@@ -134,4 +135,59 @@ class UserGroupController(
...
+            schema = Schema(implementation = ErrorResponse::class)
+          )
+        ]
+      )
+    ]
+  )
+  fun addGroupByUserId(
+    @Parameter(description = "The userId of the user.", required = true) @PathVariable
+    userId: UUID,
+    @Parameter(description = "The group to be added to the user.", required = true) @PathVariable
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserGroupController.kt:187]: Minor - the **group code** of the group to be added to the user

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserGroupControllerIntTest.kt [hunk 2]
```diff
@@ -10,6 +10,111 @@ import uk.gov.justice.digital.hmpps.externalusersapi.integration.IntegrationTest
...
+      callGetGroups(userId = "90F930E1-2195-4AFD-92CE-0EB5672DA02C")
+        .jsonPath("[?(@.groupCode == 'PECS_DRB8')]")
+        .doesNotExist()
+
+      webTestClient
+        .put().uri("/users/90F930E1-2195-4AFD-92CE-0EB5672DA02C/groups/PECS_DRB8")
+        .headers(setAuthorisation("AUTH_GROUP_MANAGER", listOf("ROLE_AUTH_GROUP_MANAGER")))
+        .exchange()
+        .expectStatus().isBadRequest
+
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserGroupControllerIntTest.kt:84]: Should we also test the message in the json returned?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserGroupControllerIntTest.kt [hunk 3]
```diff
@@ -10,6 +10,111 @@ import uk.gov.justice.digital.hmpps.externalusersapi.integration.IntegrationTest
...
+        .exchange()
+        .expectStatus().isNoContent
+
+      callGetGroups(userId = "7CA04ED7-8275-45B2-AFB4-4FF51432D1EC")
+        .jsonPath("[?(@.groupCode == 'SITE_1_GROUP_2')]")
+        .isEqualTo(mapOf("groupCode" to "SITE_1_GROUP_2", "groupName" to "Site 1 - Group 2"))
+    }
+
+    @Test
+    fun `add group to a user as group manager`() {
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserGroupControllerIntTest.kt:57]: Should we test add group to a user that doesn't exist?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt [hunk 4]
```diff
@@ -228,6 +231,148 @@ class UserGroupServiceTest {
     }
   }
 
+  @Nested
+  inner class AddGroupByUserId {
+
+    @Test
+    fun blankGroupCode() {
+      whenever(userRepository.findById(org.mockito.kotlin.any())).thenReturn(Optional.of(createSampleUser(username = "user")))
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt:239]: should be able to just use "any()"

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt [hunk 5]
```diff
@@ -228,6 +231,148 @@ class UserGroupServiceTest {
...
+      val roleLicence = Authority("ROLE_LICENCE_VARY", "Role Licence Vary")
+      val roleJoe = Authority("JOE", "Role Joe")
+      group.assignableRoles.addAll(
+        setOf(
+          GroupAssignableRole(roleLicence, group, true),
+          GroupAssignableRole(roleJoe, group, false)
+        )
+      )
+      whenever(groupRepository.findByGroupCode(anyString())).thenReturn(group)
+      doThrow(MaintainUserCheck.UserGroupRelationshipException("user", "User not with your groups")).whenever(maintainUserCheck)
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt:362]: Can just use UserGroupRelationshipException

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt [hunk 6]
```diff
@@ -228,6 +231,148 @@ class UserGroupServiceTest {
...
+      group.assignableRoles.addAll(
+        setOf(
+          GroupAssignableRole(roleLicence, group, true),
+          GroupAssignableRole(roleJoe, group, false)
+        )
+      )
+      whenever(groupRepository.findByGroupCode(anyString())).thenReturn(group)
+      service.addGroupByUserId(UUID.fromString("00000000-aaaa-0000-aaaa-0a0a0a0a0a0a"), "GROUP_LICENCE_VARY")
+      assertThat(user.groups).extracting<String> { it.groupCode }.containsOnly("GROUP_JOE", "GROUP_LICENCE_VARY")
+      assertThat(user.authorities).extracting<String> { it.roleCode }.containsOnly("LICENCE_VARY")
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt:285]: Should test the telemetry event here - and same for successAsGroupManager

File: src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserGroupController.kt (hunk 1 above)
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserGroupController.kt:187]: Done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserGroupControllerIntTest.kt (hunk 2 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserGroupControllerIntTest.kt:84]: Added

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserGroupControllerIntTest.kt (hunk 3 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserGroupControllerIntTest.kt:57]: A user that doesn't exist currently gives a 200 response. The service implementation is very similar to removing a group from a user, which I raised a ticket on. Reluctant to get into changing end point behaviour at this point. I'll add looking into this the ticket.

Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/resource/UserGroupControllerIntTest.kt:57]: I've expanded the scope of HAAR-944 to include this

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt (hunk 4 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt:239]: Changed

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt (hunk 5 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt:362]: Updated

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt (hunk 6 above)
Comment (AlTurner-MOJ) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt:289]: Added

File: src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt [hunk 7]
```diff
@@ -311,6 +319,9 @@ class UserGroupServiceTest {
       service.addGroupByUserId(UUID.fromString("00000000-aaaa-0000-aaaa-0a0a0a0a0a0a"), "GROUP_LICENCE_VARY")
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserGroupServiceTest.kt:319]: No biggie but you could have reused your userId here

**Review by karenmillermoj — APPROVED** [review]
lgtm

**Pull request merged**